File: src/main/java/se/augustoccesar/querybuilder/query/Aggregation.java
package se.augustoccesar.querybuilder.query;

import se.augustoccesar.querybuilder.builders.Buildable;
import se.augustoccesar.querybuilder.constants.CommonStrings;

/**
 * Author: augustoccesar
 * Date: 15/05/17
 */
public class Aggregation implements Buildable {
    enum Type {
        AVERAGE(" AVG "),
        COUNT(" COUNT "),
        MAX(" MAX "),
        MIN(" MIN "),
        SUM(" SUM ");

        private String value;

        Type(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    public static Type AVERAGE = Type.AVERAGE;
    public static Type COUNT = Type.COUNT;
    public static Type MAX = Type.MAX;
    public static Type MIN = Type.MIN;
    public static Type SUM = Type.SUM;

    // Attributes

    private Type type;
    private Column column;
    private String alias;

    // Constructors

    private Aggregation(String markedColumn, Type type) {
        this.type = type;
        this.column = Column.fromMarkdown(markedColumn);
        this.alias = this.type.value.toLowerCase().replaceAll("\\s", "") + "_" + this.column.getAlias();
    }

    private Aggregation(String markedColumn, Type type, String alias) {
        this.type = type;
        this.column = Column.fromMarkdown(markedColumn);
        this.alias = alias;
    }

    // Methods

    public static Aggregation average(String markedColumn) {
        return new Aggregation(markedColumn, Type.AVERAGE);
    }

    public static Aggregation average(String markedColumn, String alias) {
        return new Aggregation(markedColumn, Type.AVERAGE, alias);
    }

    public static Aggregation count(String markedColumn) {
        return new Aggregation(markedColumn, Type.COUNT);
    }

    public static Aggregation count(String markedColumn, String alias) {
        return new Aggregation(markedColumn, Type.COUNT, alias);
    }

    public static Aggregation max(String markedColumn) {
        return new Aggregation(markedColumn, Type.MAX);
    }

    public static Aggregation max(String markedColumn, String alias) {
        return new Aggregation(markedColumn, Type.MAX, alias);
    }

    public static Aggregation min(String markedColumn) {
        return new Aggregation(markedColumn, Type.MIN);
    }

    public static Aggregation min(String markedColumn, String alias) {
        return new Aggregation(markedColumn, Type.MIN, alias);
    }

    public static Aggregation sum(String markedColumn) {
        return new Aggregation(markedColumn, Type.SUM);
    }

    public static Aggregation sum(String markedColumn, String alias) {
        return new Aggregation(markedColumn, Type.SUM, alias);
    }

    // Overrides

    @Override
    public String build() {
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder
                .append(" ")
                .append(this.type.getValue())
                .append(CommonStrings.OPEN_PARENTHESES)
                .append(this.column.getPrefix()).append(".").append(this.column.getName())
                .append(CommonStrings.CLOSE_PARENTHESES)
                .append(CommonStrings.AS)
                .append(this.alias)
                .append(" ");

        return stringBuilder.toString().replaceAll("\\s+", " ");
    }

    // Getters and Setters

    public Type getType() {
        return type;
    }

    public void setType(Type type) {
        this.type = type;
    }

    public Column getColumn() {
        return column;
    }

    public void setColumn(Column column) {
        this.column = column;
    }

    public String getAlias() {
        return alias;
    }

    public void setAlias(String alias) {
        this.alias = alias;
    }
}

File: src/main/java/br/com/augustoccesar/querybuilder/query/Aggregation.kt
package br.com.augustoccesar.querybuilder.query

import br.com.augustoccesar.querybuilder.builders.Buildable
import br.com.augustoccesar.querybuilder.constants.CommonStrings

/**
 * Author: augustoccesar
 * Date: 15/05/17
 */
class Aggregation : Buildable {
    enum class Type constructor(val value: String) {
        AVERAGE(" AVG "),
        COUNT(" COUNT "),
        MAX(" MAX "),
        MIN(" MIN "),
        SUM(" SUM ")
    }

    var type: Type? = null
    var column: Column? = null
    var alias: String? = null


    private constructor(markedColumn: String, type: Type) {
        this.type = type
        this.column = Column.fromMarkdown(markedColumn)
        this.alias = this.type!!.value.toLowerCase().replace("\\s".toRegex(), "") + "_" + this.column!!.alias
    }

    private constructor(markedColumn: String, type: Type, alias: String) {
        this.type = type
        this.column = Column.fromMarkdown(markedColumn)
        this.alias = alias
    }

    override fun build(): String {
        val stringBuilder = StringBuilder()
        stringBuilder
                .append(" ")
                .append(this.type!!.value)
                .append(CommonStrings.OPEN_PARENTHESES)
                .append(this.column!!.prefix).append(".").append(this.column!!.name)
                .append(CommonStrings.CLOSE_PARENTHESES)
                .append(CommonStrings.AS)
                .append(this.alias)
                .append(" ")

        return stringBuilder.toString().replace("\\s+".toRegex(), " ")
    }

    companion object {
        @JvmField
        var AVERAGE = Type.AVERAGE
        @JvmField
        var COUNT = Type.COUNT
        @JvmField
        var MAX = Type.MAX
        @JvmField
        var MIN = Type.MIN
        @JvmField
        var SUM = Type.SUM

        @JvmStatic
        fun average(markedColumn: String): Aggregation {
            return Aggregation(markedColumn, Type.AVERAGE)
        }

        @JvmStatic
        fun average(markedColumn: String, alias: String): Aggregation {
            return Aggregation(markedColumn, Type.AVERAGE, alias)
        }

        @JvmStatic
        fun count(markedColumn: String): Aggregation {
            return Aggregation(markedColumn, Type.COUNT)
        }

        @JvmStatic
        fun count(markedColumn: String, alias: String): Aggregation {
            return Aggregation(markedColumn, Type.COUNT, alias)
        }

        @JvmStatic
        fun max(markedColumn: String): Aggregation {
            return Aggregation(markedColumn, Type.MAX)
        }

        @JvmStatic
        fun max(markedColumn: String, alias: String): Aggregation {
            return Aggregation(markedColumn, Type.MAX, alias)
        }

        @JvmStatic
        fun min(markedColumn: String): Aggregation {
            return Aggregation(markedColumn, Type.MIN)
        }

        @JvmStatic
        fun min(markedColumn: String, alias: String): Aggregation {
            return Aggregation(markedColumn, Type.MIN, alias)
        }

        @JvmStatic
        fun sum(markedColumn: String): Aggregation {
            return Aggregation(markedColumn, Type.SUM)
        }

        @JvmStatic
        fun sum(markedColumn: String, alias: String): Aggregation {
            return Aggregation(markedColumn, Type.SUM, alias)
        }
    }
}
